### ISSUE TITLE
Grammar Exclusivity Manager

### ISSUE THREAD

**Issue by LexiconCode** [dictation-toolbox/Caster#738]
**Is your feature request related to a problem? Please describe.**
Credit goes to @davidlehub for initial creation of idea and implementation thus far.

Dragonflies grammar exclusivity allows grammars to be exclusive over normal grammars. Exclusive grammars are only recognized.

This allows for a few things.  
- Exclusive Grammars can override dragons built in grammars.
- Hopefully makes creating different modes like spelling mode and such easier to implement.

**Describe the solution you'd like**
A system to manage grammar exclusivity. To what extent an exact implementation yet to be determined.

Preferably implemented in casters `Hooks and Events` system for minimal impact on core code.

**Describe alternatives you've considered**
The only alternative is to disable all of the rules and even then dragons built in rules interfere.

Summarized quotes and resources from davidlehub from gitter.

davidlehub 
> I putted Exclusivity Manager in 1 file. (for people to test)
1.Download this "init.py" from https://drive.google.com/open?id=1BD8JojzPaROke_N9ATApBt74QC3bTaUv
2. Put that file in Caster folder ("Documents\Caster")

davidlehub 
> Relate to "hooks", i suspect the "HookRunner" could be interesting. Maybe this could help: [pictures](https://drive.google.com/open?id=14Crr7BhV5vwxJHYm08NmNEPuUNLuPgzE)

davidlehub 
>To make that system more useful. I had to:
Make it works per app: Auto add "context app" rules to be exclusive, when foreground windows changed. Store/remember witch rules are been exclusives per app.
Enable Dragon vocabularies when needed.
etc.

**Comment by davidlehub** [comment]
thks

**Comment by davidlehub** [comment]
:)
Also, maybe the videos? It could help peoples to have some idea how exclusiveness can be useful.

**Comment by LexiconCode** [comment]
@davidlehub That would be a good idea. You can even update your current post with the videos.

**Comment by davidlehub** [comment]
Here they are:
https://www.youtube.com/playlist?list=PLQ6G9K9v8sUR44kvrit9pFoSBvGPKgM-7

**Comment by lettersandnumbersgithub** [comment]
I found the following in `\Dragonfly\dragonfly\grammar\grammar_base.py`
Perhaps it could be used to enable/disable exclusive mode when switching between windows/applications?
```

    def enter_context(self):
        """
            Enter context callback.

            This method is called when a phrase-start has been
            detected.  It is only called if this grammar's
            context previously did not match but now does
            match positively.

        """

    def exit_context(self):
        """
            Exit context callback.

            This method is called when a phrase-start has been
            detected.  It is only called if this grammar's
            context previously did match but now doesn't
            match positively anymore.

        """

```


Also, If all Dragon commands are disabled when in exclusive mode, perhaps we should make a list of, and devise workarounds/pass-throughs for, any Dragon commands we wish to keep? Such as:
"Go To Sleep"
"Wake Up" (if also excluded)
"Click <buttontext>"
"Press <modifiers> <key>"
 If Mimic() won't work in exclusive mode?

**Comment by davidlehub** [comment]
> I found the following in `\Dragonfly\dragonfly\grammar\grammar_base.py`
> Perhaps it could be used to enable/disable exclusive mode when switching between windows/applications?
> 
> ```
> 
>     def enter_context(self):
>         """
>             Enter context callback.
> 
>             This method is called when a phrase-start has been
>             detected.  It is only called if this grammar's
>             context previously did not match but now does
>             match positively.
> 
>         """
> 
>     def exit_context(self):
>         """
>             Exit context callback.
> 
>             This method is called when a phrase-start has been
>             detected.  It is only called if this grammar's
>             context previously did match but now doesn't
>             match positively anymore.
> 
>         """
> ```
> 
> Also, If all Dragon commands are disabled when in exclusive mode, perhaps we should make a list of, and devise workarounds/pass-throughs for, any Dragon commands we wish to keep? Such as:
> "Go To Sleep"
> "Wake Up" (if also excluded)
> "Click "
> "Press "
> If Mimic() won't work in exclusive mode?

Relate to that, there is also:
`    def _process_begin(self, executable, title, handle):
        """
            Start of phrase callback.

            *This usually is the method which should be overridden
            to give derived grammar classes custom behavior.*

            This method is called when the speech recognition
            engine detects that the user has begun to speak a
            phrase. This method is called by the
            ``Grammar.process_begin`` method only if this
            grammar's context matches positively.

            Arguments:
             - *executable* -- the full path to the module whose
               window is currently in the foreground.
             - *title* -- window title of the foreground window.
             - *handle* -- window handle to the foreground window.

        """`

**Example**, this is what i did initially to detect >> foreground windows changed :
`class grammExclusivenessCtrler_rule(MergeRule):
    # mapping = {
    # }
    
    def _process_begin(self):
        # ...
        
        activeWinHnd = win32gui.GetForegroundWindow()
        if activeWinHnd != gl.previousWinHnd:

            """ Forground windows changed, so:
            Make (or restore) a set of rules/grammars to be exclusive for the current forground window.
            """

            # ...
            
            gl.previousWinHnd = activeWinHnd`

note the following:
	- That "_process_begin" function is called ONLY when the microphone start to dectecting some things (e.g., the user start to speak in to the microphone).
	  So that means, it not detect "foreground window changed" as soon it happens. And that could cause a little latency if the process, of making the related rules/grammars to be exclusive, takes time...

(That’s why, since i don’t know how with python, i used an external tool, Autohotkey, to detect the "foreground window changed" EVENT, then tell Caster it happens.  )

**Comment by davidlehub** [comment]
> "Go To Sleep"
> "Wake Up" (if also excluded)
> "Click "
> "Press "

They are all excluded (all Dragon commands). 

(Yeaa, our Dragon is not happy and turns a deaf ear, bcz we excluded him :-D )

As Solution, what i did:
- hmm, i think is more easy to explain with a screen view... i will post a vid

**Comment by lettersandnumbersgithub** [comment]
> In short:

    Store the "state" of the current exclusiveness.
    Disable exclusiveness of all grammar. It makes Dragon commands available.
    Do the "mimic()" of a Dragon command.
    Finally, switch back to the state stored at step 1.

I was thinking the same thing :) But was wondering how quickly exclusivity can be switched on and off?

Each time a rule's exclusiveness is changed, must it be unloaded and reloaded? Or disabled and re-enabled?


If listening for ShellMessages (HSHELL_WINDOWACTIVATED, HSHELL_RUDEAPPACTIVATED, HSHELL_WINDOWCREATED especially) is not possible with Caster, one possible way to switch context when the window changes would be to check the window handle on_end(), so that any time the window changed as a result of a voice command, exclusivity could be changed post voice command and be less impactful, and would only need to be changed on_begin() the times it was changed by keyboard or mouse or an application popped up/stole focus.

Or perhaps Caster could have its own accompanying Python script which runs on its own thread and is entirely separate, much like an AutoHotkey script, which would not be bound by threading issues and could listen for ShellMessages and notify caster to change context/exclusiveness and perform numerous other functions (pun intended :p) asynchronously, perhaps it would help with GUIs also :)

**Comment by davidlehub** [comment]
> I was thinking the same thing :) But was wondering how quickly exclusivity can be switched on and off?

With the new caster system, i seems pretty fast. But event with the old one, when it has a delay (bcz of waiting the grammars to become exclusive), It doesn't that bad.
But anyway, i think, is preferable to wait a little bit (worst case 2 to 3 sec) and have Caster respond more accurately,  then been frustrated by no needed vocabularies interfering and execute wrong commands.

**Comment by davidlehub** [comment]
> Each time a rule's exclusiveness is changed, must it be unloaded and reloaded? Or disabled and re-enabled?

If i am not wrong, it disabled and re-enabled, except for the grammar/s of the CCR...

Pretty sure is the same with the new Caster system.

**Comment by davidlehub** [comment]
> Or perhaps Caster could have its own accompanying Python script which runs on its own thread and is entirely separate, much like an AutoHotkey script, which would not be bound by threading issues and could listen for ShellMessages and notify caster to change context/exclusiveness and perform numerous other functions (pun intended :p) asynchronously, perhaps it would help with GUIs also :)

Oh yea, agreed :)

**Comment by davidlehub** [comment]
> > "Go To Sleep"
> > "Wake Up" (**if also excluded**)
> > "Click "
> > "Press "
> 
> They are all excluded (all Dragon commands).
> 
> (Yeaa, our Dragon is not happy and turns a deaf ear, bcz we excluded him :-D )
> 
> As Solution, what i did:
> 
> * hmm, i think is more easy to explain with a screen view... i will post a video.
> 
> In short:
> 
> 1. Store the "state" of the current exclusiveness.
> 2. Disable exclusiveness of all grammar. It makes Dragon commands available.
> 3. Do the "mimic()" of a Dragon command.
> 4. Finally, switch back to the state stored at step 1.

So, to use a Dragon command, for example "Go To Sleep", here is a "summary" of what i did (in the old Caster system):
https://drive.google.com/open?id=160GLTsiEu3XDfVr-iCtDVgTecqAAvWkf

**Comment by LexiconCode** [comment]
> - That "_process_begin" function is called ONLY when the microphone start to dectecting some things (e.g., the user start to speak in to the microphone).

I think it should be fine though because that's how dragonfly the text window changes for all grammars anyway and I haven't seen really any performance issues with it.

**Comment by Danesprite** [comment]
@LexiconCode Thanks for passing on my messages :+1:

It would be pretty useful to have an exclusivity manager for enabling/disabling Dragon's built in rules like this!

It would be possible to implement a `Rule.set_exclusiveness()` method for setting Dragonfly rules as exclusive/non-exclusive. It would be quite tricky to implement though and couldn't really work well with non-exported rules, which is how CCR usually works.

**Comment by lettersandnumbersgithub** [comment]
_"Go To Sleep"_
"Wake Up" (if also excluded)
"Click [button]"
"Press [modifier] [key]"

"Go To Sleep" is also blocked by my free dictation interception `"<text>":` ,  so I was able to devise and test this now, despite not yet having the ability to enable exclusiveness.
The following are working replacements:

```
import natlink
```
```
"Go To Sleep":         Function(lambda: natlink.setMicState("sleeping") ),
"Mic Off":         Function(lambda: natlink.setMicState("off") ),
```

As we cannot get the dictation while the microphone is asleep (inconvenient, but good for security and privacy), if "Wake Up": is also blocked, perhaps the only solution will be to disable all exclusiveness when going to sleep and re-enable it again upon waking (easily done with changeCallback())

**Comment by LexiconCode** [comment]
> It would be quite tricky to implement though and couldn't really work well with non-exported rules, which is how CCR usually works.

The way the current implementation works around this is disabling all rules then re-enabling exclusive only rules. There's not much added benefit to this outside of blocking Dragon.

I think the first step implementation here would be to enable exclusivity for all Caster grammars.

Going beyond that to work with CCR is going to be tricky. Honestly beyond my ability to pull off but I could imagine it something like this.

1. Disabling the merged grammars and backing up current enabled grammars.
2. Run any CCR exclusive grammars through merger with a unique exclusive prefix(so it doesn't conflict).
3. Once done with the exclusive mode/grammars, delete their unique merges and restore by enabling the previous date mentioned in step one.

If the system worked as described there wouldn't be the penalty of reemerging everything in the first step when exiting exclusivity it would be just re-enabled

**Comment by davidlehub** [comment]
> If the system worked as described there wouldn't be the penalty of reemerging everything in the first step when exiting exclusivity it would be just re-enabled.

:thumbsup: No reemerging = No delay, especially when the callback "_process_begin" is use to detect "foreground window" changes.

**Comment by LexiconCode** [comment]
@davidlehub 
For making exclusivity for all loaded Caster grammars. Simply placing `grammar.set_exclusiveness(1)` right before grammar.load() works.

To be placed in the following lines [MappingRules](https://github.com/dictation-toolbox/Caster/blob/5531ddb8bd4ead8f861a26c0e266a359f93ae0fb/castervoice/lib/ctrl/mgr/grammar_manager.py#L260) and [MergeRule](
https://github.com/dictation-toolbox/Caster/blob/5531ddb8bd4ead8f861a26c0e266a359f93ae0fb/castervoice/lib/ctrl/mgr/grammar_manager.py#L274)

Now that we know where to place the `event` we can implement a `hook` to do more advanced functionality but it is a basic proof of concept.

I'm a bit short on time @lettersandnumbersgithub could fill you in on some of the other issues like Dragon `sleeping` and so forth.

**Comment by lettersandnumbersgithub** [comment]
>  For making exclusivity for all loaded Caster grammars. Simply placing `grammar.set_exclusiveness(1)` right before grammar.load() as a proof of concept.

>  For making exclusivity for all loaded Caster grammars. Simply placing `grammar.set_exclusiveness(1)` right **after** grammar.load() as a proof of concept.

:)

**Comment by lettersandnumbersgithub** [comment]
With this change alone, grammars remain exclusive and active even after putting Dragon to sleep
(for which the workaround I described above is required)
However I have written the following which appears to solve this problem:
```
"Go To Sleep":         Function(lambda: GoToSleep() ),
```

```
def GoToSleep():
   for grammar in get_engine().grammars:
      grammar.set_exclusiveness(0)
   natlink.setMicState("sleeping")
```

**Comment by LexiconCode** [comment]
@davidlehub 
@lettersandnumbersgithub 
All right I have a working `hook` an `event` for exclusivity for all _Caster_ loaded grammars. Note this does not include Dragonfly grammars as they are not loaded through Casters Grammar Manager.

[see the related code](https://github.com/dictation-toolbox/Caster/compare/master...LexiconCode:exclusive_hook#diff-5788216756570170f4c1a8984eff8ea4R277)

If you try out the code above you'll need to start dns once. Then go to `hooks.toml` and set `ExclusiveHook = true` then restart DNS again.

I'll take some feedback and questions here before making a pull request.

**Comment by davidlehub** [comment]
> With this change alone, grammars remain exclusive and active even after putting Dragon to sleep
> (for which the workaround I described above is required)
> However I have written the following which appears to solve this problem:
> 
> ```
> "Go To Sleep":         Function(lambda: GoToSleep() ),
> ```
> 
> ```
> def GoToSleep():
>    for grammar in get_engine().grammars:
>       grammar.set_exclusiveness(0)
>    natlink.setMicState("sleeping")
> ```

Just some think relate to it, on my mind now: in the function "GoToSleep", maybe storing the "state" of the current exclusiveness (for example, put in a list, all grammars, or rules, been exclusive). So when Dragon is "wake up", we restore the saved state.

Code would be some thing like:
`def GoToSleep():
   --> Store/save Things Currently Been Exclusive
   for grammar in get_engine().grammars:
      grammar.set_exclusiveness(0)
   natlink.setMicState("sleeping")`

and then when, for example:
`def WakeUp():
    --> reStore Things that was been Exclusive
    ....
`

:)

**Comment by davidlehub** [comment]
@LexiconCode 

> If you try out the code above you'll need to start dns once. Then go to hooks.toml and set ExclusiveHook = true or say enable exclusive hook then restart DNS again.

how to try it? i click on "Create pull request"?

**Comment by davidlehub** [comment]
![image](https://user-images.githubusercontent.com/47902139/73906447-f0b2ad00-4870-11ea-8d44-5c0e5c2b28b4.png)

**Comment by LexiconCode** [comment]
> how to try it? i click on "Create pull request"?
> ![image](https://user-images.githubusercontent.com/47902139/73906463-fd370580-4870-11ea-8afe-a114c230482c.png)

There's a few ways to go about doing this. 
Git - you'll need to add my repository has a remote. Once that's done you can check out the branches. See [Git Extensions](https://git-extensions-documentation.readthedocs.io/en/latest/remote_feature.html) Docs

Zip - Download my branch through the zip file by going to my [exclusive branch](https://github.com/LexiconCode/Caster/tree/exclusive_hook) in my caster fork.

**Comment by davidlehub** [comment]
have error: **AttributeError: 'NoneType' object has no attribute '__file__'**
```
Caster: castervoice is up-to-date
Error loading _caster from C:\Users\HP\Documents\Caster\_caster.py
Traceback (most recent call last):
  File "C:\NatLink\NatLink\MacroSystem\core\natlinkmain.py", line 331, in loadFile
    imp.load_module(modName,fndFile,fndName,fndDesc)
  File "C:\Users\HP\Documents\Caster\_caster.py", line 65, in <module>
    control.init_nexus(_content_loader)
  File "C:\Users\HP\Documents\Caster\castervoice\lib\control.py", line 7, in init_nexus
    _NEXUS = Nexus(content_loader)
  File "C:\Users\HP\Documents\Caster\castervoice\lib\ctrl\nexus.py", line 83, in __init__
    self._load_and_register_all_content(rules_config, hooks_runner, transformers_runner)
  File "C:\Users\HP\Documents\Caster\castervoice\lib\ctrl\nexus.py", line 92, in _load_and_register_all_content
    content = self._content_loader.load_everything(rules_config)
  File "C:\Users\HP\Documents\Caster\castervoice\lib\ctrl\mgr\loading\load\content_loader.py", line 57, in load_everything
    rules = self._process_requests(rule_requests)
  File "C:\Users\HP\Documents\Caster\castervoice\lib\ctrl\mgr\loading\load\content_loader.py", line 106, in _process_requests
    content_item = self.idem_import_module(request.module_name, request.content_type)
  File "C:\Users\HP\Documents\Caster\castervoice\lib\ctrl\mgr\loading\load\content_loader.py", line 92, in idem_import_module
    return fn()
  File "C:\Users\HP\Documents\Caster\castervoice\rules\core\utility_rules\window_mgmt_rule.py", line 20, in get_rule
  File "C:\Users\HP\Documents\Caster\castervoice\lib\ctrl\mgr\rule_details.py", line 31, in __init__
    self._filepath = RuleDetails._calculate_filepath_from_frame(stack, 1)
  File "C:\Users\HP\Documents\Caster\castervoice\lib\ctrl\mgr\rule_details.py", line 37, in _calculate_filepath_from_frame
    filepath = module.__file__.replace("\\", "/")
AttributeError: 'NoneType' object has no attribute '__file__'
-- skip unchanged wrong grammar file: C:\Users\HP\Documents\Caster\_caster.p
```

**Comment by LexiconCode** [comment]
> have error: **AttributeError: 'NoneType' object has no attribute '**file**'**
delete window_mgmt_rule.pyc

**Comment by davidlehub** [comment]
[+] Mannulaly make "ExclusiveHook = true", in the hooks.toml, exclusiveness works as expected.

[-] When speech "disable exclusive hook", the spec is recognised, but nothing change: exclusiveness is still actived/enabled. In hooks.toml, "ExclusiveHook " still = true. Even after restart Dragon.

[-] After manually change  "ExclusiveHook = false", in the hooks.toml,  then  speech "enable exclusive hook", the spec is recognised, but nothing change: exclusiveness is still deactived/disabled. "ExclusiveHook " still = false. Even after restart Dragon.

**Comment by LexiconCode** [comment]
Okay that's pretty much what I expected.

**Comment by davidlehub** [comment]
Hi guys :)
I wonder if there is already a function (or place in the code) that detect foreground/active window changed?

**Comment by davidlehub** [comment]
what i mean is a LOGIC that detect it. 
Not the methodes, like get_active_window_title , in castervoice\lib\utilities.py

**Comment by LexiconCode** [comment]
@davidlehub you mean what makes the method through dragonfly?


I don't think there's a way through the engine to expose when there's a context change.

**Comment by davidlehub** [comment]
how can i use/refer this function `def _change_rule_enabled(self, class_name, enabled, tail=True)`
(witch is in '\castervoice\lib\ctrl\mgr\grammar_manager.py'), other way then `_NEXUS._grammar_manager._change_rule_enabled(iRclassName, False)`?

**Comment by davidlehub** [comment]
> @davidlehub you mean what makes the method through dragonfly?
> 
> I don't think there's a way through the engine to expose when there's a context change.

OK, then i am gonna try to create one. In fact gonna use a method i used on old old caster :)

**Comment by LexiconCode** [comment]
> 
> 
> how can i use/refer this function `def _change_rule_enabled(self, class_name, enabled, tail=True)`
> (witch is in '\castervoice\lib\ctrl\mgr\grammar_manager.py'), other way then `_NEXUS._grammar_manager._change_rule_enabled(iRclassName, False)`?

Yes you can use this to the event system. The printer hooked makes use of this.

**Comment by davidlehub** [comment]
Ahh, supper!

**Comment by davidlehub** [comment]
Can we have 2 hooks with a same returned value of `def get_pronunciation(self):`?

for example, having 2 hooks with same pronuncation "A":
`
class hook1(BaseHook):

	...

	def get_pronunciation(self):
		return "A"


class hook2(BaseHook):

	...

	def get_pronunciation(self):
		return "A"
`

**Comment by LexiconCode** [comment]
> 
> 
> Can we have 2 hooks with a same returned value of `def get_pronunciation(self):`?
> 
> for example, having 2 hooks with same pronuncation "A":
> `
> class hook1(BaseHook):
> 
> ```
> ...
> 
> def get_pronunciation(self):
> 	return "A"
> ```
> 
> class hook2(BaseHook):
> 
> ```
> ...
> 
> def get_pronunciation(self):
> 	return "A"
> ```
> 
> `

No that is not possible as think of it like a grammar in the sense of need its unique Pronunciation.

**Comment by davidlehub** [comment]
Hehe, little weird thing just happen with about an GUI:

A GUI button is auto click when i switch windows/apps! 

The funny is, this weird thing could be useful for detecting a foreground window has changed. But for now, the button pressed event should be fired only with clicked with a mouse.

To see that, check  repository: branch 'NewExclusiveness':
- run with Run_Caster_DNS.bat
- Watch for the log "Gui button 'Simulate NewAppDetected' pressed.". 

Also try, when the second button ("click to On/Off") is pressed, then switch to an other window, the first button is auto focused and pressed. Event if the Gui is minimized.

**Comment by LexiconCode** [comment]
That's a creative workaround! However I think the best way to proceed is to create a function that monitors the foreground window in dragonfly. Initially this would be first for Windows and then other implementations specific to each OS.

This [article](https://devblogs.microsoft.com/oldnewthing/20131202-00/?p=2503) has some low-level windows information that can be used to create a function to monitor processes. it lays out four steps to create this method. Fortunately we only need to be concerned with the first as the rest are already implemented as dragonfly.

**Comment by davidlehub** [comment]
I think i won't use that GUI, for now, bcz i don't have the control on it (button should only press when i click on it, not auto click...).

**Comment by davidlehub** [comment]
but thanks :) >That's a creative workaround!
>  to create a function that monitors the foreground window in dragonfly. 

hmm, since i don't know how to to it in "dragonfly", for now i think i'm gonna try to do it(the detection of foreground changed) where it detect "Speech start detected." (logged message):
- by comparing previous and current windowID, ...

**Comment by davidlehub** [comment]
Hey @LexiconCode, again: things are modifying in my code , like was happen in utility.py. Probably by accidentally, but... that never happen strangely like that before.

Funny... :)

**Comment by LexiconCode** [comment]
@davidlehub keep working on what you're doing and I'll see what I can come up with.

**Comment by davidlehub** [comment]
About Git PUSH, what about the files in my "AppData\Local\caster": are they upload to github too? I guess is Yes.

**Comment by LexiconCode** [comment]
> 
> 
> About Git PUSH, what about the files in my "AppData\Local\caster": are they upload to github too? I guess is Yes.

nope

**Comment by davidlehub** [comment]
Hi guys, is the order of elements in the list '_enabled_ordered' (in side rules.toml) important? Or it can be in any order, like elements of a set()?

**Comment by LexiconCode** [comment]
The enabled ordered is actually important. Editing that field directly is not recommended by hand or programmatically (outside of letting Caster manage it).

**Comment by davidlehub** [comment]
So in case you guys wanna try/test the current progress of the exclusiveness thing:
- branch: https://github.com/davidlehub/Caster/tree/NewExclusiveness
- Edit this file 'C:\Python27\Lib\site-packages\dragonfly\__main__.py':
--- inside the `def _on_begin():` function, add: 
`
    from castervoice.exclusiveness.Notify_on_begin_fromDragonFly import Notify_on_begin_fromDragonFly
    Notify_on_begin_fromDragonFly()
`
--- So it looks like this:
`
def _on_begin():
    print("Speech start detected.")

    from castervoice.exclusiveness.Notify_on_begin_fromDragonFly import Notify_on_begin_fromDragonFly
    Notify_on_begin_fromDragonFly()
`

Note: My first main goal is to make things works. Very probably, the codes are not well structure, optimized, not implementing the new hooks-event system, etc. I adapt my codes from old caster version.
:)

**Comment by davidlehub** [comment]
> The enabled ordered is actually important. Editing that field directly is not recommended by hand or programmatically (outside of letting Caster manage it).

ho oh... hmm, will take care of this later later. bcz for now my exclusiveness system affect that variable (_enabled_ordered).

**Comment by LexiconCode** [comment]
That should be okay as I believe that's what internally Caster uses to manage that variable.

**Comment by davidlehub** [comment]
> That should be okay as I b

oof. 
Still, if it causes problem, i will have to modify and use list, instead of set() ...

**Comment by davidlehub** [comment]
Yeaa, quick test and it works: Rules related to the foreground window are set to be exclusive. For exemple: switch to Visual sTudio Code then say "match bracket". Then switch to Google Chrome then say "new Tab". 
So, for each app/window, only "their" rules are available. Plus, some rules that we want them to be always exclusive: see 'RulesCasterAlwayNeed_className' in 'castervoice\exclusiveness\globalVariable\GlobalV.py'
'

**Comment by davidlehub** [comment]
this is the content of my 'hooks.toml':
ExclusiveHook = false
FormattingHook = false
PrinterHook = true
RulePrinterHook = false
ShowStatusWindowOnErrorHook = true
storeAllRulesLoadedHook = true

not sure if it need to be exactly like that to work tho. I put it here, just in case.

**Comment by davidlehub** [comment]
The system if far from been completed: lots of things/feature to add, more tests. Just the basic "witch rule(s) to be available, for an foreground window" exclusively

**Comment by davidlehub** [comment]
If you have a change to try/test it, any comments are appreciated :) 

Hope there is no bugs. Cz, i hate bugs :D, jook

**Comment by LexiconCode** [comment]
Sounds good I will try to test as I can between work. Might take a number of days though.

**Comment by davidlehub** [comment]
New Update: some bug fix and optimizations

-*** Must have the following hooks:
ExclusiveHook = true
FormattingHook = false
PrinterHook = true
RulePrinterHook = false
ShowStatusWindowOnErrorHook = true
storeAll_LoadedRule_Hook = true
storeAll_mergedCcr_Hook = true
storeAll_registeredRule_hook = true

**Comment by LexiconCode** [comment]
I am working on upgrading the hooks and events system to have functions that fire when they're enabled or disabled.

**Comment by davidlehub** [comment]
IS it normal that some time it say, in the log message, a command is  recognized but no action is executed. And have to restart Dragon to make it works correctly?

**Comment by LexiconCode** [comment]
No that's not normal. Although you need to make sure something else isn't stealing focus where an action is executed but the perceived result looks like it hasn't happened. However if you're talking about the situation where a grammar is exclusive and commands are being recognized from a non-exclusive grammar that would be a separate issue.

**Comment by davidlehub** [comment]
> However if you're talking about the situation where a grammar is exclusive and commands are being recognized from a non-exclusive grammar that would be a separate issue.

No, not that situation.

That bugs happen when i restart caster only (without restarting Dragon).  And, like i said, i restart Dragon, things works.

I suspect there is some kind 'garbage' grammars that are not cleanup (delete). When we do 'grammar.set_exclusiveness(1)', and there is new grammars that are generated/remerged, that old grammars been set_exclusiveness(1) have to be cleanup...

Will investigate about that late, late, later. For now: just restart Dragon :)

**Comment by davidlehub** [comment]
New update:  save and restore states.

For example,  let says we are with VScode:
- enable python (by saying enable python)
- switch to, let says Chrome. The rules for Chrome gonna be part of exclusiveness (with other important/core rules, examples: Alphabet, number, navigation, etc). And the Vscode rules are no more exclusive. Same for for Python rules.
- Then switch back to VScode, rules that was exclusives for that window, are restored back (including the python one, in this case)

Need this to work:

> -*** Must have the following hooks:
ExclusiveHook = true
FormattingHook = false
PrinterHook = true
RulePrinterHook = false
ShowStatusWindowOnErrorHook = true
storeAll_LoadedRule_Hook = true
storeAll_mergedCcr_Hook = true
storeAll_registeredRule_hook = true

and:
So in case you guys wanna try/test the current progress of the exclusiveness thing:

branch: https://github.com/davidlehub/Caster/tree/NewExclusiveness
Edit this file 'C:\Python27\Lib\site-packages\dragonfly__main__.py':
--- inside the def _on_begin(): function, add:
`
from castervoice.exclusiveness.Notify_on_begin_fromDragonFly import Notify_on_begin_fromDragonFly
Notify_on_begin_fromDragonFly()
`
--- So it looks like this:

`
def _on_begin():
print("Speech start detected.")

from castervoice.exclusiveness.Notify_on_begin_fromDragonFly import Notify_on_begin_fromDragonFly
Notify_on_begin_fromDragonFly()
`

**Comment by davidlehub** [comment]
> Hehe, little weird thing just happen with an GUI wx:
> 
> A GUI button is auto click when i switch windows/apps!
> 
> The funny is, this weird thing could be useful for detecting a foreground window has changed. But for now, the button pressed event should be fired only with clicked with a mouse.
> 
> To see that, check repository: branch 'NewExclusiveness', (the gui is at the end of "_caster.py")
> 
> * run with Run_Caster_DNS.bat
> * Watch for the log "Gui button 'Simulate NewAppDetected' pressed.".
> 
> Also try, when the second button ("click to On/Off") is pressed, then switch to an other window, the first button is auto focused and pressed. Event if the Gui is minimized.

Found the reason: is bcz of my Autohotkey that interact with it...

**Comment by LexiconCode** [comment]
`Add HUD` pull request should solve a number of issues for you.
- It keeps track of active CCR and Mapping Rules (it doesn't control them just tracks them). more like what the engines these rather than Caster. WIP
- Implements recognition observers so you don't have to utilize Notify_on_begin_fromDragonFly()` in dragonfly, [see here](https://github.com/dmakarov/Caster/blob/4142b85fbbeedb117c860029f596222aff2f10f4/_caster.py#L30)`

**Comment by davidlehub** [comment]
New Update: "Exclusive Uniq Mode"  feature: done implemetented.

The Uniq Mode: is a kind of multi layer of exclusiveness. Could be used in situation where we need to restrain the vocabularies.

For now, i don't have a good used case example yet. Maybe we won't need it at all. But is there, bcz  i just re-implementing features from  my old caster.

**Comment by davidlehub** [comment]
New Update: an example of 'Exclusive Uniq Mode'.

- Code location:  castervoice\exclusiveness\EXAMPLES\Exclusive_UniqMode_test.py
- Say: "test uniq mode"
- Result: Only spec/commands of  ALphabet and Number rules are available.
- To end/finish that Uniq Mode, say: "finish" or "done"

Also, see the file 'castervoice\exclusiveness\globalVariable\ExclusivenessSetting.py' for:
- how to customize rules for exclusiveness.

**Comment by davidlehub** [comment]
And also an example of: turning OFF exclusiveness, then using a DNS command, then switch back to exclusiveness.

- Code location: castervoice\exclusiveness\EXAMPLES\Dragon_Controler.py
- Say: "micof"
- Result: DNS microphone is turning off.

**Comment by davidlehub** [comment]
For now, i think all the exclusiveness 'system' is there.

Of curse there is, i think, lots of things to improve (code structure, optimization, etc.) to make it more suitable/compatible, user friendly for non programmer users. And Since my knowledge of python and programming is not that advance, i won't spend time in it. So go ahead, do what ever you want with the code. And maybe, there is others approach/way to isolate DNS vocabularies :)

**Comment by LexiconCode** [comment]
Thank you for all the effort you put into this. and some of the other features we talked about previously that would make this easier are merged we will reevaluate how it's implemented.

**Comment by davidlehub** [comment]
Happy that i can contribute :)

But again, big thanks to you, are all others ppls, for making Caster more and more powerful. Cz, can't believe at 2020 we still using fingers, almost only, to controls computers. hihi :)

**Comment by davidlehub** [comment]
Update: Fixing some behavior when turning on/off the microphone by voice (makeDragonHeard)

**Comment by davidlehub** [comment]
how to use 'class Observer'? [https://github.com/dmakarov/Caster/blob/4142b85fbbeedb117c860029f596222aff2f10f4/_caster.py#L30](url)

**Comment by LexiconCode** [comment]
```python
class Observer(RecognitionObserver):
    def __init__(self):
        self.hud = control.nexus().comm.get_com("hud")

    def on_begin(self):
        pass

    def on_recognition(self, words):
        try:
            self.hud.send("$ {}".format(" ".join(words)))
        except ConnectionRefusedError:  # pylint: disable=undefined-variable
            print("$ {}".format(" ".join(words)))

    def on_failure(self):
        try:
            self.hud.send("?!")
        except ConnectionRefusedError:  # pylint: disable=undefined-variable
            print("?!")

Observer().register()  
```

**Comment by davidlehub** [comment]
1- i append that code to _caster.py?
2- or i create a new file 'observer.py', and put it, let say, in 'exclusiveness' folder?

**Comment by davidlehub** [comment]
i'm completely lost: no idea what is mean:
>Add HUD pull request should solve a number of issues for you.

**Comment by davidlehub** [comment]
UPdate: new way to see exclusiveness system: in fact, that system what it does is it **exclude** speech engine, vocabulary (or rules?). So i think is more clear and less confuse for end user, to call 'speech engine exclusion' (instead of 'exclusive' ...).
- So, I create a mapping rule called: 'speechEngineExclusion'. Then to enable or disable that 'feature', user just say, as spec: 'enable speech engine exclusion'. Then:
- The speech engine vocabulary is exclude: only active caster's rules are available.
- (to include back, say 'enable ...', then speech engine vocab is available with caster's rules).
Ahh, also, as consequence: no need to use hook (so 'exclusiveHok' can be deleted).  So 1 less thing to remember to activate/deactivate foe end user.
And also: no more 'haking' (hihi) in to dragon fly file. Replace by _process_begin() in SpeedEngineExclusion.py .   

Some main/interesting files (in directory and subdir 'castervoice\exclusiveness':
- SpeedEngineExclusion.py
- ExclusivenessSetting.py
- DragonVocabulary_cls.py  (could be renamed to speechEngineVocabulary?)
- the files in folder 'EXAMPLES'
